### ISSUE TITLE
[CT-807] post_model_hook Ignores snowflake_warehouse config

### ISSUE THREAD

**Issue by ribolzisalvador** [dbt-labs/dbt-snowflake#187]
### Describe the bug
When building or running a model with the Snowflake adapter the snowflake_warehouse config defined in dbt_project is ignored in favour of the warehouse defined in the used profile, but only in the [post_model_hook](https://github.com/dbt-labs/dbt-snowflake/blob/main/dbt/adapters/snowflake/impl.py#L85)

### Steps To Reproduce
In `dbt_project` : 

```
models:
    project:
        +snowflake_warehouse: "TRANSFORM_LARGE_WH"
```

In profile:
```
dev:
  target: dev-large
  outputs:
    dev-large:
      type: snowflake
      warehouse: TRANSFORM_DEV_WH
```

### Expected behavior
Expected behavior would be to use only the warehouse defined at the config level if it exists, and not only the default warehouse

### Screenshots and log output

I modified `impl.py` to log what warehouse was being used and printed it like so:

```
    def post_model_hook(self, config: Mapping[str, Any], context: Optional[str]) -> None:
        default_warehouse = self.config.credentials.warehouse
        warehouse = config.get("snowflake_warehouse", default_warehouse)
        print(f"Post | Default Warehouse: {default_warehouse}, Config Warehouse: {warehouse}")
        print(f"Context: {context}")
        if context is not None:
            self._use_warehouse(context)
```

> 13:44:21.959658 [debug] [Thread-51 ]: Began executing node model.ussbi.clean_zu_account
> 13:44:21.960571 [debug] [Thread-51 ]: Using snowflake connection "model.ussbi.clean_zu_account"
> 13:44:21.960704 [debug] [Thread-51 ]: On model.ussbi.clean_zu_account: /* {"app": "dbt", "dbt_version": "1.1.1", "profile_name": "dev", "target_name": "dev-large", "node_id": "model.ussbi.clean_zu_account"} */
> select current_warehouse() as warehouse
> 13:44:21.960830 [debug] [Thread-51 ]: Opening a new connection, currently in state closed
> 13:44:23.626242 [debug] [Thread-51 ]: SQL status: SUCCESS 1 in 1.66 seconds
> 13:44:23.628425 [debug] [Thread-51 ]: Using snowflake connection "model.ussbi.clean_zu_account"
> 13:44:23.628591 [debug] [Thread-51 ]: On model.ussbi.clean_zu_account: /* {"app": "dbt", "dbt_version": "1.1.1", "profile_name": "dev", "target_name": "dev-large", "node_id": "model.ussbi.clean_zu_account"} */
> **use warehouse TRANSFORM_LARGE_WH**
> 13:44:23.866643 [debug] [Thread-51 ]: SQL status: SUCCESS 1 in 0.24 seconds
> 13:44:23.883869 [debug] [Thread-51 ]: Writing runtime SQL for node "model.ussbi.clean_zu_account"
> 13:44:23.885314 [debug] [Thread-51 ]: Using snowflake connection "model.ussbi.clean_zu_account"
> 13:44:23.885460 [debug] [Thread-51 ]: On model.ussbi.clean_zu_account: /* {"app": "dbt", "dbt_version": "1.1.1", "profile_name": "dev", "target_name": "dev-large", "node_id": "model.ussbi.clean_zu_account"} */
> 
> {Model SQL redacted}
> 
> 13:44:24.397210 [debug] [Thread-51 ]: SQL status: SUCCESS 1 in 0.51 seconds
> **Post | Default Warehouse: TRANSFORM_DEV_WH, Config Warehouse: TRANSFORM_LARGE_WH** 
> **Context: TRANSFORM_DEV_WH**
> 13:44:24.403195 [debug] [Thread-51 ]: Using snowflake connection "model.ussbi.clean_zu_account"
> 13:44:24.403345 [debug] [Thread-51 ]: On model.ussbi.clean_zu_account: /* {"app": "dbt", "dbt_version": "1.1.1", "profile_name": "dev", "target_name": "dev-large", "node_id": "model.ussbi.clean_zu_account"} */
> **use warehouse TRANSFORM_DEV_WH**
> 13:44:24.656715 [debug] [Thread-51 ]: SQL status: SUCCESS 1 in 0.25 seconds
> 13:44:24.658586 [debug] [Thread-51 ]: finished collecting timing info
> 13:44:24.658798 [debug] [Thread-51 ]: On model.ussbi.clean_zu_account: Close
> 

### System information
**The output of `dbt --version`:**
```
Core:
  - installed: 1.1.1
  - latest:    1.1.1 - Up to date!

Plugins:
  - snowflake: 1.1.0 - Up to date!
```

Also tried with dbt==1.0.8

**The operating system you're using:**
MacOS 12.2.1

**The output of `python --version`:**
Python 3.9.13

### Additional context

I believe replacing the `self._use_warehouse(context)` call with something like the pre_model_hook 
```
        warehouse = config.get("snowflake_warehouse", default_warehouse)
        if warehouse == default_warehouse or warehouse is None:
            return None
        previous = self._get_warehouse()
        self._use_warehouse(warehouse)
```
should fix this.

Thanks!

**Comment by jtcohen6** [comment]
@ribolzisalvador Thanks for opening! I'm not sure I totally follow, so I've asked a few clarifying questions down below.

The goal of the `post_model_hook` step — which, unlike the name implies, is _not_ to run user-provided `post-hook` for this model — is really just to enact the `snowflake_warehouse` config. In particular, the `pre_model_hook` switches from the default warehouse (defined in `profiles.yml`) to the model-specific warehouse (as configured by `snowflake_warehouse`), and then the `post_model_hook` resets back to the default warehouse — in both cases, by running `use warehouse {warehouse}` commands.

Do I understand your proposal to be: Instead of using the `warehouse` from `profiles.yml` to define the default warehouse, used in the `post_model_hook` "reset," dbt should instead look to the config in `dbt_project.yml` for that default?

That is, following your example:
```sql
-- models/specific_model.sql

{{ config(snowflake_warehouse = 'MODEL_SPECIFIC_WAREHOUSE') }}

select ...
```
```yml
# dbt_project.yml -- project-level default
models:
    project:
        +snowflake_warehouse: PROJECT_DEFAULT_WAREHOUSE
```
In profile:
```yml
# profiles.yml -- connection default
dev:
  target: dev-large
  outputs:
    dev-large:
      type: snowflake
      warehouse: PROFILE_DEFAULT_WAREHOUSE
```
Is your thought that, after `specific_model` runs with the `MODEL_SPECIFIC_WAREHOUSE`, dbt should be "resetting" to `PROJECT_DEFAULT_WAREHOUSE` instead of `PROFILE_DEFAULT_WAREHOUSE`?

I don't think that would work with the way project resources are configured in dbt today, but I want to make sure I understand the behavior you're seeing as a bug!

**Comment by ribolzisalvador** [comment]
Hi @jtcohen6 ! Thanks for the quick reply,

Our main goal is to be able to have different Virtual Warehouses in Snowflake, and then select which one we want to use dynamically within DBT Cloud on a per-job basis (for the most part). 

So say we have a couple of very large jobs to run in DBT and we want them to run on a Large warehouse, instead of the default Medium warehouse all jobs would run in. 
The only way we saw of doing this was by changing the target on the DBT Cloud Job, and then have configs that check for the target and set the warehouse to be used, and that's how we arrived to the `+snowflake_warehouse` config.
We do not use nor intend to use specific_model with MODEL_SPECIFIC_WAREHOUSE at this time.

Context aside, I definitely misread the purpose of `post_model_hook` , now I see it really intends to go back to the PROFILE_DEFAULT_WAREHOUSE rather than PROJECT_DEFAULT_WAREHOUSE. This _can_ error out if there is no `warehouse` defined in the profiles.yml, since the post_hook would issue a `use warehouse None;`

I feel like we are good to move forward using this snowflake_warehouse config in dbt cloud then and this is intended behaviour not a bug, just out of curiosity why revert to the PROFILE_DEFAULT_WAREHOUSE instead of PROJECT_DEFAULT_WAREHOUSE?
